### PR TITLE
Fix stats hijack

### LINKED ISSUE #1315: websocket: Unable to hijack the connection
`v1.2.0-rc2`
`v1.2.0`

etcd backend on Debian box

Configuration (read from etcd):
```
etcdctl get /traefik/backends/r-announcer/servers/server41/url
ws://172.25.255.41:2010

etcdctl get /traefik/frontends/r-announcer/passHostHeader
false

etcdctl get /traefik/frontends/r-announcer/backend
r-announcer

etcdctl get /traefik/frontends/r-announcer/entrypoints
https

etcdctl get /traefik/frontends/r-announcer/routes/host/rule
Host:announcer.mol.net.ua

etcdctl get /traefik/entrypoints/https/address
:443

etcdctl get /traefik/entrypoints/https/compress
false

tls settngs skipped because they work with https wihout any problems

```

```
time="2017-03-22T21:05:47+03:00" level=debug msg="Creating frontend r-announcer" 
time="2017-03-22T21:05:47+03:00" level=debug msg="Wiring frontend r-announcer to entryPoint http" 
time="2017-03-22T21:05:47+03:00" level=debug msg="Creating route host Host:announcer.mol.net.ua" 
time="2017-03-22T21:05:47+03:00" level=debug msg="Wiring frontend r-announcer to entryPoint https" 
time="2017-03-22T21:05:47+03:00" level=debug msg="Creating route host Host:announcer.mol.net.ua" 
time="2017-03-22T21:05:47+03:00" level=debug msg="Creating backend r-announcer" 
time="2017-03-22T21:05:47+03:00" level=debug msg="Creating load-balancer wrr" 
time="2017-03-22T21:05:47+03:00" level=debug msg="Creating server server41 at ws://172.25.255.41:2010 with weight 0"
```

The settings for proxiyng web sockets are fairly standard. I have studied thoroughly https://github.com/containous/traefik/issues/452, so there are no:

ddr balancer (on this backend)
circuitbreaker (on this backend)
retry (globally)
compress (on entrypoint https)

But all log is filled with:

> Unable to hijack the connection: *negroni.responseWriter the ResponseWriter doesn't support the Hijacker interface

and all connections to frontend URL immediately return 500 without any attempts to call backend.

### PR DESCRIPTION
**Pull request by emilevauge**
### Description

This PR fixes #1315 and adds a generic http recover.

Fixes #1315, #1591

### CODE REVIEW

File: middlewares/recover.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"github.com/codegangsta/negroni"
+	"github.com/containous/traefik/log"
+	"net/http"
+)
+
+// RecoverHandler recovers from a panic in http handlers
+func RecoverHandler(next http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
```
Comment (marco-jantke) [inline middlewares/recover.go]: You could extract this anonymous fn and reuse it in the NegroniRecoverHandler.

File: middlewares/stateful.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package middlewares
+
+import "net/http"
+
+// Stateful interface groups all http interface that must be
```
Comment (marco-jantke) [inline middlewares/stateful.go]: Typo `HTTP interfaces ..`.

File: server/server.go [hunk 3]
```diff
@@ -525,6 +525,12 @@ func (server *Server) prepareServer(entryPointName string, router *middlewares.H
 		return nil, err
 	}
 
+	middlewaresStr := ""
+	for _, handler := range negroni.Handlers() {
+		middlewaresStr += reflect.TypeOf(handler).String() + " "
```
Comment (marco-jantke) [inline server/server.go]: Is this here by intend or a left over?

Comment (emilevauge) [inline server/server.go]: I added this intentionally ;)

File: middlewares/stateful.go (hunk 2 above)
Comment (emilevauge) [inline middlewares/stateful.go]: fixed

File: middlewares/recover.go (hunk 1 above)
Comment (emilevauge) [inline middlewares/recover.go]: fixed

File: server/server.go (hunk 3 above)
Comment (marco-jantke) [inline server/server.go]: Ok :) To be honest I am not sure how useful I find this. With a simple configuration I currently get the following output:
```
Middlewares: negroni.HandlerFunc *middlewares.Logger *stats.Stats negroni.HandlerFunc
```
Additionally I have for example retries configured. It is not mentioned here, as this middleware is setup in the `loadConfig()` method and specific to a backend. Therefore we should maybe make this message more explicit or extend the logging also on other positions. WDYT?

Comment (emilevauge) [inline server/server.go]: Yep, you are right, as we don't have middlewares created in `loadConfig`, it is a bit useless. I will remove it.

**Comment by Regner** [comment]
Any chance of this helping with #1591?

File: middlewares/recover.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package middlewares
+
+import (
+	"github.com/codegangsta/negroni"
+	"github.com/containous/traefik/log"
+	"net/http"
```
Comment (ldez) [inline middlewares/recover.go]: Please group the non-stdlib imports below the stdlib ones 😉

File: middlewares/recover_test.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package middlewares
+
+import (
+	"github.com/codegangsta/negroni"
+	"net/http"
+	"net/http/httptest"
+	"testing"
```
Comment (ldez) [inline middlewares/recover_test.go]: Please group the non-stdlib imports below the stdlib ones.

**Comment by emilevauge** [comment]
@Regner Indeed, it will fix the error `Unable to hijack the connection: *middlewares.responseRecorder` :)

File: middlewares/recover.go (hunk 4 above)
Comment (emilevauge) [inline middlewares/recover.go]: Ouch, I finally configured Gogland correctly on imports ^^

**Comment by Regner** [comment]
Thanks @emilevauge. I wasn't entirely sure but if thats the case I eagerly look forward to this being merged.

File: middlewares/recover_test.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	recoverHandler := RecoverHandler(http.HandlerFunc(fn))
+	server := httptest.NewServer(recoverHandler)
+	defer server.Close()
+
+	resp, err := http.Get(server.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != 500 {
+		t.Fatalf("Received non-%d response: %d\n", 500, resp.StatusCode)
```
Comment (timoreimann) [inline middlewares/recover_test.go]: We don't really need to replace the 500 error code with a variable, do we?

File: middlewares/recover_test.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	}
+}
+
+func TestNegroniRecoverHandler(t *testing.T) {
```
Comment (timoreimann) [inline middlewares/recover_test.go]: I think the two tests could be fairly easily merged into a single table-driven test by letting the handler be created through a `func() http.Handler` or so?

Comment (emilevauge) [inline middlewares/recover_test.go]: @timoreimann to be honest, as we are in the process of removing `negroni` dep, I don't think that's what we want. `TestNegroniRecoverHandler` will be removed soon ;)

File: middlewares/recover_test.go (hunk 6 above)
Comment (emilevauge) [inline middlewares/recover_test.go]: We do ^^

File: middlewares/recover_test.go (hunk 7 above)
Comment (timoreimann) [inline middlewares/recover_test.go]: Yeah good point. 👍

**Pull request merged**